src/libraries/System.Private.CoreLib/src/System/Span.cs

```diff
@@ -413,8 +421,17 @@ public Span<T> Slice(int start, int length)
             // of this is that if either input is negative or if the input sum overflows past Int32.MaxValue,
             // that information is captured correctly in the comparison against the backing _length field.
             // We don't use this same mechanism in a 32-bit process due to the overhead of 64-bit arithmetic.
-            if ((ulong)(uint)start + (ulong)(uint)length > (ulong)(uint)_length)
-                ThrowHelper.ThrowArgumentOutOfRangeException();
+            // We special-case "start == 0" to optimize codegen since the pattern Slice(0, ...) is so common.
+            if (RuntimeHelpers.IsKnownConstant(start) && start == 0)
+            {
+                if ((uint)length > (uint)_length)
+                    ThrowHelper.ThrowArgumentOutOfRangeException();
+            }
```

Comment:
Any reason the above path shouldn't apply to 32-bit as well?

Reply:
As far as I can tell, this is already optimized correctly in 32-bit scenarios because the JIT performs the correct transforms.

```cs
if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
if ((uint)0 > (uint)_length || (uint)length > (uint)(_length - 0))
if (false /* _length known to be non-negative */ || (uint)length > (uint)_length)
if ((uint)length > (uint)_length)
```

Unless I'm completely misreading the x86 codegen. Which is entirely possible. :)